alpha/template/semver/semver.go

```diff
@@ -272,88 +270,84 @@ func (sv *semverTemplate) generateChannels(semverChannels *bundleVersions) []dec
-			if curSkips.Has(finalEntry.Replaces) {
-				finalEntry.Skips = curSkips.Difference(sets.NewString(finalEntry.Replaces)).List()
-			} else {
-				finalEntry.Skips = curSkips.List()
-			}
-		}
-
-		if archChange || kindChange || xChange {
-			// we don't maintain skips/replaces over these transitions
-			curSkips = sets.NewString()
-			prevZMax = ""
-		} else {
-			if yChange {
-				prevZMax = prevTuple.name
+	// bundle --> version lookup
+	bundleVersions := make(map[string]semver.Version)
+	for _, vs := range *harvestedVersions {
+		for b, v := range vs {
+			if _, ok := bundleVersions[b]; !ok {
+				bundleVersions[b] = v
 			}
-			curSkips.Insert(prevTuple.name)
 		}
 	}
 
-	// last entry accumulation
-	lastTuple := entries[len(entries)-1]
-	prevChannel := unlinkedChannels[lastTuple.parent]
-	finalEntry := &prevChannel.Entries[lastTuple.index]
-	finalEntry.Replaces = prevZMax
-	// don't include replaces in skips list, but they are accumulated in discrete cycles (and maybe useful for later channels) so remove here
-	if curSkips.Has(finalEntry.Replaces) {
-		finalEntry.Skips = curSkips.Difference(sets.NewString(finalEntry.Replaces)).List()
-	} else {
-		finalEntry.Skips = curSkips.List()
-	}
+	for _, channel := range unlinkedChannels {
+		entries := &channel.Entries
+		sort.Slice(*entries, func(i, j int) bool {
+			return bundleVersions[(*entries)[i].Name].LT(bundleVersions[(*entries)[j].Name])
+		})
 
-	for _, ch := range unlinkedChannels {
-		channels = append(channels, *ch)
+		// "inchworm" through the sorted entries, iterating curEdge but extending yProbe to the next Y-transition
+		// then catch up curEdge to yProbe as 'skips', and repeat until we reach the end of the entries
+		// finally, because the inchworm will always fail to pick up the last Y-transition, we test for it and link it up as a 'replaces'
+		curEdge, yProbe := 0, 0
+		zmaxQueue := ""
+		entryCount := len(*entries)
+
+		for curEdge < entryCount {
+			for yProbe < entryCount {
+				curVersion := bundleVersions[(*entries)[curEdge].Name]
+				yProbeVersion := bundleVersions[(*entries)[yProbe].Name]
+				if getMinorVersion(yProbeVersion).EQ(getMinorVersion(curVersion)) {
+					yProbe += 1
+				} else {
+					break
+				}
+			}
+			// if yProbe crossed a threshold, the previous entry is the last of the previous Y-stream
+			preChangeIndex := yProbe - 1
+
+			if curEdge != yProbe {
+				if zmaxQueue != "" {
+					// add skips edge to allow skipping over y iterations within an x stream
+					(*entries)[preChangeIndex].Skips = append((*entries)[preChangeIndex].Skips, zmaxQueue)
+					(*entries)[preChangeIndex].Replaces = zmaxQueue
+				}
+				zmaxQueue = (*entries)[preChangeIndex].Name
+			}
+			for curEdge < preChangeIndex {
+				// add skips edges to y-1 from z < y
+				(*entries)[preChangeIndex].Skips = append((*entries)[preChangeIndex].Skips, (*entries)[curEdge].Name)
+				curEdge += 1
+			}
+			curEdge += 1
+			yProbe = curEdge + 1
+		}
```

Comment:
So after spending more time dissecting this for loop to really understand what each step is doing (took ~30min) I think it is good to go logically. Maybe it's just me coming back from being sick and still being a bit foggy (still recovering) but I felt that it was a bit difficult to understand this change in what I felt was a reasonable time frame (~20 min) as a reviewer, which led me to feeling rather unconfident in my review and thus withholding an "approval".

Feel free to take this recommendation or leave it as it, but I think a slight refactor for readability could be useful here. Maybe something along the lines of:
```suggestion
			yStreamHead := getYStreamHead(curEntry, *entries) // This function would perform the same logic as already exists in the yProbe for loop 

			if curEdge != yStreamHead {
				if zmaxQueue != "" {
					// add skips edge to allow skipping over y iterations within an x stream
					(*entries)[yStreamHead].Skips = append((*entries)[yStreamHead].Skips, zmaxQueue)
					(*entries)[yStreamHead].Replaces = zmaxQueue
				}
				zmaxQueue = (*entries)[yStreamHead].Name
			}
			
			// could also be made into a separate function if desired
			for _, edge := range (*entries)[curEdge:yStreamHead] {
				// add skips edges to y-1 from z < y
				(*entries)[yStreamHead].Skips = append((*entries)[yStreamHead].Skips, edge.Name)
			}
			curEdge = yStreamHead + 1
		}
```

Reply:
Overall, what threw me for a loop (pun intended) the most was wrapping my head around the multiple curEdge increments (and why) as we progressed through the loop and the initially unclear discovery of the y stream head (although looking back at it it totally makes sense as is, so feel free to keep the loop instead of breaking out to another function if you want - I just ate lunch so my mind is feeling a bit more clear now 😆 ).